Code/Query/NullQueryAlgebra.h

```diff
@@ -0,0 +1,106 @@
+//
+//
+//  Copyright (C) 2019 Greg Landrum and T5 Informatics GmbH
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#ifndef _RD_NULLQUERYALGEBRA_H
+#define _RD_NULLQUERYALGEBRA_H
+
+#include <GraphMol/QueryOps.h>
+
+namespace RDKit {
+namespace {
+template <class T>
+void composeBothNullQ(T *&returnQuery, T *&otherNullQ,
+                      Queries::CompositeQueryType how) {
+  bool negatedQ = returnQuery->getNegation();
+  bool negatedOtherQ = otherNullQ->getNegation();
+
+  if (how == Queries::COMPOSITE_AND) {
+    // This is the only case in which we need to do anything
+    if (!negatedQ && negatedOtherQ) {
+      returnQuery->setNegation(true);
+    }
+  } else if (how == Queries::COMPOSITE_OR) {
+    // This is the only case in which we need to do anything
+    if (negatedQ && !negatedOtherQ) {
+      returnQuery->setNegation(false);
+    }
+  } else if (how == Queries::COMPOSITE_XOR) {
+    if (!negatedQ && !negatedOtherQ) {
+      returnQuery->setNegation(true);
+    } else if (negatedQ + negatedOtherQ == 1) {
+      returnQuery->setNegation(false);
+    }
+  }
+}
+
+template <class T>
+void composeNullQFirst(T *&returnQuery, T *&otherQ,
```

Comment:
Not your fault at all, but I spent an embarassingly long amount of time with a piece of paper convincing myself the logic here is correct.
negated null queries are hard to think about. ;-)

Reply:
I don't even think we use negated null queries at any point... but I wanted to make sure that we supported them, in case any popped up at some point.

Any way, if you think about it, it is quite simple:

- Null queries and negated null queries are opposite extremes: null queries always match, negates ones, never match (they are an absolute negation).

- In AND combinations the narrower condition prevails: null queries always lose, negated null queries always stay.

- OR combinations are the opposite: we keep the wider condition: null queries always win, negated ones always lose.

- XOR queries are more complicated, though: against a null query, we keep the negation of the other query. With negated null queries, we directly keep the other query.